src/folding/nova/ivc.rs

```diff
@@ -0,0 +1,279 @@
+        // assign them directly to w_i, u_i.
+        self.w_i = Witness::<C1>::new(w_i1.clone(), self.r1cs.A.n_rows);
+        self.u_i = self.w_i.commit(&self.pedersen_params, vec![u_i1_x])?;
+
+        // set values for next iteration
+        self.i += C1::ScalarField::one();
+        self.z_i = z_i1.clone();
+        self.U_i = U_i1.clone();
+        self.W_i = W_i1.clone();
+
+        Ok(())
+    }
+
+    /// Implements IVC.V
+    pub fn verify(&mut self, z_0: Vec<C1::ScalarField>, num_steps: u32) -> Result<(), Error> {
+        if self.i != C1::ScalarField::from(num_steps) {
+            return Err(Error::IVCVerificationFail);
+        }
+
+        if self.u_i.x.len() != 1 || self.U_i.x.len() != 1 {
+            return Err(Error::IVCVerificationFail);
+        }
+
+        // check that u_i's output points to the running instance
+        // u_i.X == H(i, z_0, z_i, U_i)
+        let expected_u_i_x = self
+            .U_i
+            .hash(&self.poseidon_config, self.i, z_0, self.z_i.clone())?;
+        if expected_u_i_x != self.u_i.x[0] {
+            return Err(Error::IVCVerificationFail);
+        }
+
+        // check u_i.cmE==0, u_i.u==1 (=u_i is a un-relaxed instance)
+        if self.u_i.cmE != C1::zero() || self.u_i.u != C1::ScalarField::one() {
+            return Err(Error::IVCVerificationFail);
+        }
+
+        // check R1CS satisfiability
+        self.r1cs.check_instance_relation(&self.w_i, &self.u_i)?;
+        // check RelaxedR1CS satisfiability
+        self.r1cs
+            .check_relaxed_instance_relation(&self.W_i, &self.U_i)?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ark_pallas::{Fr, Projective};
+    use ark_vesta::Projective as Projective2;
+
+    use crate::folding::nova::circuits::tests::TestFCircuit;
+    use crate::transcript::poseidon::{tests::poseidon_test_config, PoseidonTranscript};
+
+    #[test]
+    fn test_ivc() {
+        let mut rng = ark_std::test_rng();
+        let poseidon_config = poseidon_test_config::<Fr>();
+
+        let F_circuit = TestFCircuit::<Fr> { _f: PhantomData };
+        let z_0 = vec![Fr::from(3_u32)];
+
+        let mut ivc =
+            IVC::<Projective, Projective2, TestFCircuit<Fr>, PoseidonTranscript<Projective>>::new(
+                &mut rng,
+                poseidon_config.clone(), // notice that transcript_config could be different than poseidon (eg. keccak's transcript config)
+                poseidon_config,         // poseidon config
+                F_circuit,
+                z_0.clone(),
+            )
+            .unwrap();
+
+        let num_steps: usize = 3;
+        for _ in 0..num_steps {
+            ivc.prove_step().unwrap();
+        }
+
+        ivc.r1cs
+            .check_relaxed_instance_relation(&ivc.w_i, &ivc.u_i)
+            .unwrap();
+        ivc.r1cs
+            .check_relaxed_instance_relation(&ivc.W_i, &ivc.U_i)
+            .unwrap();
```

Comment:
Seems this will also be checked in `ivc.verify`?

Reply:
Yes! good catch, thanks. I had those lines before the verify method was implemented, but after the commit that adds the verification method it does not make sense to keep them as you pointed out. Thanks for the catch! ^^